File: examples/rust/snippets/getting_started.rs
use anyhow::{Result};
use bip39::{Language, Mnemonic};
use gl_client::{
    bitcoin::Network,
    credentials::{Device, Nobody},
    node::ClnClient,
    pb::{cln, cln::{amount_or_any, Amount, AmountOrAny}},
    scheduler::Scheduler,
    signer::Signer,
};
use rand::RngCore;
use std::{env, fs, path::PathBuf};
use tokio;

const NETWORK: Network = Network::Regtest;
const TEST_NODE_DATA_DIR: &str = "/tmp/gltests/node2";

fn save_to_file(file_name: &str, data: &[u8]) -> Result<()> {
    let path = PathBuf::from(TEST_NODE_DATA_DIR).join(file_name);
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(&path, data)?;
    Ok(())
}

fn read_file(file_name: &str) -> Result<Vec<u8>> {
    let path = PathBuf::from(TEST_NODE_DATA_DIR).join(file_name);
    Ok(fs::read(path)?)
}

async fn upgrade_device_certs_to_creds(
    scheduler: &Scheduler<Nobody>,
    signer: &Signer,
    creds_path: &str,
) -> Result<Device> {
    // ---8<--- [start: upgrade_device_certs_to_creds]
    let device = Device::from_path(creds_path);
    let upgraded = device.upgrade(scheduler, signer).await?;
    save_to_file("credentials_upgraded.gfs", &upgraded.to_bytes())?;
    // ---8<--- [end: upgrade_device_certs_to_creds]
    Ok(upgraded)
}

fn create_seed() -> Result<Vec<u8>> {
    // ---8<--- [start: create_seed]
    let mut rng = rand::thread_rng();
    let mut entropy = [0u8; 32];
    rng.fill_bytes(&mut entropy);

    // Seed phrase for user
    let mnemonic = Mnemonic::from_entropy_in(Language::English, &entropy)?;
    let _phrase = mnemonic.words().collect::<Vec<_>>().join(" ");

    const EMPTY_PASSPHRASE: &str = "";
    let seed = &mnemonic.to_seed(EMPTY_PASSPHRASE)[0..32]; // Only need the first 32 bytes

    // Store the seed on the filesystem, or secure configuration system
    save_to_file("hsm_secret", seed)?;
    // ---8<--- [end: create_seed]
    Ok(seed.to_vec())
}

fn load_developer_creds() -> Result<Nobody> {
    // ---8<--- [start: dev_creds]
    let developer_cert_path = env::var("GL_NOBODY_CRT")?;
    let developer_key_path = env::var("GL_NOBODY_KEY")?;
    let developer_cert = std::fs::read(developer_cert_path).unwrap_or_default();
    let developer_key = std::fs::read(developer_key_path).unwrap_or_default();
    let developer_creds = Nobody {
        cert: developer_cert,
        key: developer_key,
        ..Nobody::default()
    };
    // ---8<--- [end: dev_creds]
    Ok(developer_creds)
}

async fn register_node(seed: Vec<u8>, developer_creds: Nobody) -> Result<(Scheduler<Nobody>, Device, Signer)> {
    // ---8<--- [start: init_signer]
    let signer = Signer::new(seed.clone(), NETWORK, developer_creds.clone())?;
    // ---8<--- [end: init_signer]

    // ---8<--- [start: register_node]
    let scheduler = Scheduler::new(NETWORK, developer_creds).await?;

    // Passing in the signer is required because the client needs to prove
    // ownership of the `node_id`
    let registration_response = scheduler.register(&signer, None).await?;
    // ---8<--- [start: device_creds]
    let device_creds = Device::from_bytes(registration_response.creds);
    save_to_file("credentials.gfs", &device_creds.to_bytes())?;
    // ---8<--- [end: device_creds]
    // ---8<--- [end: register_node]
    Ok((scheduler, device_creds, signer))
}

async fn get_node(scheduler: &Scheduler<Device>) -> Result<ClnClient> {
    // ---8<--- [start: get_node]
    let node = scheduler.node().await?;
    // ---8<--- [end: get_node]
    Ok(node)
}

async fn start_node(device_creds_file_path: &str) -> Result<(cln::GetinfoResponse, cln::ListpeersResponse, cln::InvoiceResponse)> {
    // ---8<--- [start: start_node]
    let creds = Device::from_path(device_creds_file_path);
    let scheduler = Scheduler::new(NETWORK, creds.clone()).await?;
    let mut node: ClnClient = scheduler.node().await?;
    // ---8<--- [end: start_node]

    // ---8<--- [start: list_peers]
    let info = node.getinfo(cln::GetinfoRequest::default()).await?;
    let info = info.into_inner();
    let peers = node.list_peers(cln::ListpeersRequest::default()).await?;
    let peers = peers.into_inner();
    // ---8<--- [end: list_peers]

    // ---8<--- [start: start_signer]
    let seed = read_file("hsm_secret")?;
    let signer = Signer::new(seed, NETWORK, creds.clone())?;
    let (_tx, rx) = tokio::sync::mpsc::channel(1);
    tokio::spawn(async move {
        signer.run_forever(rx).await.unwrap();
    });
    // ---8<--- [end: start_signer]

    // ---8<--- [start: create_invoice]
    let amount = AmountOrAny {
        value: Some(amount_or_any::Value::Amount(Amount { msat: 10_000 })),
    };
    let invoice = node
        .invoice(cln::InvoiceRequest {
            amount_msat: Some(amount),
            description: format!("desc_{}", rand::random::<u32>()),
            label: format!("label_{}", rand::random::<u32>()),
            ..Default::default()
        })
        .await?;
    let invoice = invoice.into_inner();
    // ---8<--- [end: create_invoice]
    Ok((info, peers, invoice))
}

async fn recover_node(dev_creds: Nobody) -> Result<(Scheduler<Nobody>, Device, Signer)> {
    // ---8<--- [start: recover_node]
    let seed = read_file("hsm_secret")?;
    let signer = Signer::new(seed.clone(), NETWORK, dev_creds.clone())?;
    let scheduler = Scheduler::new(NETWORK, dev_creds).await?;
    let recover_response = scheduler.recover(&signer).await?;
    // ---8<--- [end: recover_node]
    let device_creds = Device::from_bytes(recover_response.creds);
    save_to_file("credentials.gfs", &device_creds.to_bytes())?;
    Ok((scheduler, device_creds, signer))
}

#[tokio::main]
async fn main() -> Result<()> {
    println!("Creating seed...");
    let seed = create_seed()?;

    println!("Loading developer credentials...");
    let developer_creds = load_developer_creds()?;

    println!("Registering node...");
    let (scheduler, device_creds, signer) = register_node(seed, developer_creds.clone()).await?;
    println!("Node Registered!");

    println!("Getting node information...");
    let device_scheduler = Scheduler::new(NETWORK, device_creds.clone()).await?;
    let _gl_node = get_node(&device_scheduler).await?;    

    let (info, peers, invoice) = start_node(&format!("{TEST_NODE_DATA_DIR}/credentials.gfs")).await?;
    println!("Node pubkey: {}", hex::encode(info.id));
    println!("Peers list: {:?}", peers.peers);
    println!("Invoice created: {}", invoice.bolt11);

    println!("Upgrading certs...");
    let _upgraded = upgrade_device_certs_to_creds(&scheduler, &signer, &format!("{TEST_NODE_DATA_DIR}/credentials.gfs")).await?;

    println!("Recovering node...");
    let (_scheduler2, _device_creds2, _signer2) = recover_node(developer_creds.clone()).await?;
    println!("Node Recovered!");

    let (info, _peers, _invoice) = start_node(&format!("{TEST_NODE_DATA_DIR}/credentials.gfs")).await?;
    println!("Node pubkey: {}", hex::encode(info.id));

    println!("All steps completed successfully!");

    Ok(())
}

File: examples/python/snippets/getting_started.py
import os
import sys
import string
from pathlib import Path
import bip39  # type: ignore
from glclient import Credentials, Signer, Scheduler  # type: ignore
from pathlib import Path
from pyln import grpc as clnpb  # type: ignore
import secrets  # Make sure to use cryptographically sound randomness

GL_NOBODY_CRT = os.environ.get('GL_NOBODY_CRT')
GL_NOBODY_KEY = os.environ.get('GL_NOBODY_KEY')
NETWORK="regtest"
TEST_NODE_DATA_DIR_1="/tmp/gltests/node1"


def save_to_file(file_name: str, data: bytes) -> None:
    file_name = Path(TEST_NODE_DATA_DIR_1) / file_name
    os.makedirs(file_name.parent, exist_ok=True)    
    with open(file_name, "wb") as file:
        file.write(data)


def read_file(file_name: str) -> bytes:
    file_name = Path(TEST_NODE_DATA_DIR_1) / file_name
    with open(file_name, "rb") as file:
        return file.read()


def upgrade_device_certs_to_creds(scheduler: Scheduler, signer: Signer, device_creds_file_path: str):
    # ---8<--- [start: upgrade_device_certs_to_creds]
    device_creds = Credentials.from_path(device_creds_file_path)
    upgraded_device_creds = device_creds.upgrade(scheduler.inner, signer.inner)
    # ---8<--- [end: upgrade_device_certs_to_creds]
    return upgraded_device_creds


def create_seed() -> bytes:
    # ---8<--- [start: create_seed]
    rand = secrets.randbits(256).to_bytes(32, "big")  # 32 bytes of randomness

    # Seed phrase for user
    phrase = bip39.encode_bytes(rand)

    seed = bip39.phrase_to_seed(phrase)[:32]  # Only need the first 32 bytes

    # Store the seed on the filesystem, or secure configuration system
    save_to_file("hsm_secret", seed)

    # ---8<--- [end: create_seed]
    return seed


def load_developer_creds():
    # ---8<--- [start: dev_creds]
    developer_cert_path = os.environ.get('GL_NOBODY_CRT')
    developer_key_path = os.environ.get('GL_NOBODY_KEY')
    developer_cert = Path(developer_cert_path).open(mode="rb").read()
    developer_key = Path(developer_key_path).open(mode="rb").read()

    developer_creds = Credentials.nobody_with(developer_cert, developer_key)
    # ---8<--- [end: dev_creds]
    return developer_creds


def register_node(seed: bytes, developer_creds: Credentials):
    # ---8<--- [start: init_signer]
    signer = Signer(seed, NETWORK, developer_creds)
    # ---8<--- [end: init_signer]

    # ---8<--- [start: register_node]
    scheduler = Scheduler(NETWORK, developer_creds)

    # Passing in the signer is required because the client needs to prove
    # ownership of the `node_id`
    registration_response = scheduler.register(signer, invite_code=None)

    # ---8<--- [start: device_creds]
    device_creds = Credentials.from_bytes(registration_response.creds)
    save_to_file("credentials.gfs", device_creds.to_bytes())
    # ---8<--- [end: device_creds]
    # ---8<--- [end: register_node]
    return scheduler, device_creds, signer


def get_node(scheduler: Scheduler, device_creds: Credentials) -> dict:
    # ---8<--- [start: get_node]
    scheduler = scheduler.authenticate(device_creds)
    node = scheduler.node()
    # ---8<--- [end: get_node]
    return node


def start_node(device_creds_file_path: str) -> None:
    # ---8<--- [start: start_node]
    device_creds = Credentials.from_path(device_creds_file_path)
    scheduler = Scheduler(NETWORK, device_creds)

    node = scheduler.node()
    # ---8<--- [end: start_node]

    # ---8<--- [start: list_peers]
    getinfo_response = node.get_info()
    listpeers_response = node.list_peers()
    # ---8<--- [end: list_peers]

    # ---8<--- [start: start_signer]
    seed = read_file("hsm_secret")
    signer = Signer(seed, NETWORK, device_creds)

    signer.run_in_thread()
    # ---8<--- [end: start_signer]

    # ---8<--- [start: create_invoice]
    invoice_response = node.invoice(
        amount_msat=clnpb.AmountOrAny(amount=clnpb.Amount(msat=10000)),
        description="description".join(secrets.choice(string.ascii_letters + string.digits) for _ in range(10)),
        label="label".join(secrets.choice(string.ascii_letters + string.digits) for _ in range(10)),
    )
    # ---8<--- [end: create_invoice]
    return getinfo_response, listpeers_response, invoice_response


def recover_node(developer_cert: bytes, developer_key: bytes) -> None:
    # ---8<--- [start: recover_node]
    seed = read_file("hsm_secret")
    signer_creds = Credentials.nobody_with(developer_cert, developer_key)
    signer = Signer(seed, NETWORK, signer_creds)
    scheduler = Scheduler(NETWORK, signer_creds)
    recover_response = scheduler.recover(signer)
    # ---8<--- [end: recover_node]
    device_creds = Credentials.from_bytes(recover_response.creds)
    save_to_file("credentials.gfs", device_creds.to_bytes())
    return scheduler, device_creds, signer


def main():
    # Verify files exist
    if not GL_NOBODY_CRT or not os.path.exists(GL_NOBODY_CRT):
        print(f"Error: Developer certificate not found at {GL_NOBODY_CRT}")
        sys.exit(1)
        
    if not GL_NOBODY_KEY or not os.path.exists(GL_NOBODY_KEY):
        print(f"Error: Developer key not found at {GL_NOBODY_KEY}")
        sys.exit(1)
    
    #Create seed
    print("Creating seed...")
    seed = create_seed()
    
    print("Loading developer credentials...")
    developer_creds = load_developer_creds()

    # Register node
    print("Registering node...")
    scheduler, device_creds, signer = register_node(seed, developer_creds)
    print("Node Registered!")

    # Get GL node
    print("Getting node information...")
    get_node(scheduler, device_creds)
    
    # Print node's information to check
    getinfo_response, listpeers_response, invoice_response = start_node(TEST_NODE_DATA_DIR_1 + "/credentials.gfs")
    print("Node pubkey:", getinfo_response.id.hex())
    print("Peers list:", listpeers_response.peers)
    print("Invoice created:", invoice_response.bolt11)

    # Upgrade Certificates
    print("Upgrading certs...")
    upgrade_device_certs_to_creds(scheduler, signer, TEST_NODE_DATA_DIR_1 + "/credentials.gfs")
    
    # Recover the node
    print("Recovering node...")
    scheduler, device_creds, signer = recover_node(Path(GL_NOBODY_CRT).open(mode="rb").read(), Path(GL_NOBODY_KEY).open(mode="rb").read())
    print("Node Recovered!")

    # Print node's information to check
    getinfo_response, listpeers_response, invoice_response = start_node(TEST_NODE_DATA_DIR_1 + "/credentials.gfs")
    print("Node pubkey:", getinfo_response.id.hex())
    print("All steps completed successfully!")


if __name__ == "__main__":
    main()
